cmd/kubeadm/app/phases/certs/certlist_test.go

```diff
@@ -209,6 +214,312 @@ func TestCreateCertificateChain(t *testing.T) {
 
 }
 
+func TestLeafCertificates(t *testing.T) {
+	errAny := errors.New("match any error")
+
+	type testCase struct {
+		name          string
+		certificates  Certificates
+		result        Certificates
+		assertion     func(*testing.T, *testCase)
+		setup         func(*testing.T, *testCase)
+		expectedError error
+	}
+	tests := []testCase{
+		{
+			name: "success",
+		},
+		{
+			name: "nil certs",
+			setup: func(t *testing.T, tc *testCase) {
+				tc.certificates = nil
+				tc.assertion = func(t *testing.T, tc *testCase) {
+					assert.Len(t, tc.result, 0)
+				}
+			},
+		},
+		{
+			name: "no certs",
+			setup: func(t *testing.T, tc *testCase) {
+				tc.certificates = Certificates{}
+				tc.assertion = func(t *testing.T, tc *testCase) {
+					assert.Len(t, tc.result, 0)
+				}
+			},
+		},
+		{
+			name: "error unknown ca",
+			setup: func(t *testing.T, tc *testCase) {
+				tc.certificates = Certificates{
+					{Name: "cert1", CAName: "ca1"},
+				}
+			},
+			expectedError: errAny,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			test.certificates = Certificates{
+				{Name: "ca1"},
+				{Name: "ca1-cert1", CAName: "ca1"},
+				{Name: "ca1-cert2", CAName: "ca1"},
+				{Name: "ca2"},
+				{Name: "ca2-cert1", CAName: "ca2"},
+			}
+
+			if test.assertion == nil {
+				test.assertion = func(t *testing.T, tc *testCase) {
+					assert.Len(t, tc.result, 3)
```

Comment:
I should add more assertions here, to check that only non-CA certificates are returned, for example.